protocol/grpc/write_message.go

```diff
@@ -0,0 +1,208 @@
+	_ binding.BinaryWriter     = (*pbEventWriter)(nil)
+)
+
+func (b *pbEventWriter) SetStructuredEvent(ctx context.Context, f format.Format, event io.Reader) error {
+	if b.Attributes == nil {
+		b.Attributes = make(map[string]*pb.CloudEventAttributeValue)
+	}
+
+	b.Attributes[contenttype], _ = attributeFor(f.MediaType())
+
+	var buf bytes.Buffer
+	_, err := io.Copy(&buf, event)
+	if err != nil {
+		return err
+	}
+
+	// TODO: check the data content type and set the right data format
+	b.Data = &pb.CloudEvent_BinaryData{
+		BinaryData: buf.Bytes(),
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) Start(ctx context.Context) error {
+	if b.Attributes == nil {
+		b.Attributes = make(map[string]*pb.CloudEventAttributeValue)
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) End(ctx context.Context) error {
+	return nil
+}
+
+func (b *pbEventWriter) SetData(reader io.Reader) error {
+	buf, ok := reader.(*bytes.Buffer)
+	if !ok {
+		buf = new(bytes.Buffer)
+		_, err := io.Copy(buf, reader)
+		if err != nil {
+			return err
+		}
+	}
+
+	b.Data = &pb.CloudEvent_BinaryData{
+		BinaryData: buf.Bytes(),
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) SetAttribute(attribute spec.Attribute, value interface{}) error {
+	switch attribute.Kind() {
+	case spec.SpecVersion:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid SpecVersion type, expected string got %T", value)
+		}
+		b.SpecVersion = val
+		return nil
+	case spec.ID:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid ID type, expected string got %T", value)
+		}
+		b.Id = val
+		return nil
+	case spec.Source:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid Source type, expected string got %T", value)
+		}
+		b.Source = val
+		return nil
+	case spec.Type:
+		val, ok := value.(string)
+		if !ok {
+			return fmt.Errorf("invalid Type type, expected string got %T", value)
+		}
+		b.Type = val
+		return nil
+	case spec.DataContentType:
+		if value == nil {
+			delete(b.Attributes, contenttype)
+		}
+		b.Attributes[contenttype], _ = attributeFor(value)
+		return nil
+	case spec.DataSchema:
+		if value == nil {
+			delete(b.Attributes, prefix+dataSchema)
+		}
+		b.Attributes[prefix+dataSchema], _ = attributeFor(value)
+		return nil
+	case spec.Subject:
+		if value == nil {
+			delete(b.Attributes, prefix+subject)
+		}
+		b.Attributes[prefix+subject], _ = attributeFor(value)
+		return nil
+	case spec.Time:
+		if value == nil {
+			delete(b.Attributes, prefix+time)
+		}
+		b.Attributes[prefix+time], _ = attributeFor(value)
+		return nil
+	default:
+		if value == nil {
+			delete(b.Attributes, prefix+attribute.Name())
+		}
+		b.Attributes[prefix+attribute.Name()], _ = attributeFor(value)
+	}
+
+	return nil
+}
+
+func (b *pbEventWriter) SetExtension(name string, value interface{}) (err error) {
+	if value == nil {
+		delete(b.Attributes, prefix+name)
+	}
+
+	b.Attributes[prefix+name], err = attributeFor(value)
+
+	return
```

Comment:
Let's define 'err' in the func instead of in the return arg - just for consistency

Reply:
done.